src/Nethermind/Nethermind.Shutter/ShutterTxLoader.cs

```diff
@@ -0,0 +1,242 @@
+        Metrics.KeysReceivedTimeOffset = offset;
+        string offsetText = offset < 0 ? $"{-offset}ms before" : $"{offset}ms after";
+        _logger.Info($"Got {sequencedTransactions.Count} encrypted transactions from Shutter sequencer contract for slot {keys.Slot} at time {offsetText} slot start...");
+
+        Transaction[] transactions = DecryptSequencedTransactions(sequencedTransactions, keys.Keys);
+
+        if (_logger.IsDebug && transactions.Length > 0) _logger.Debug($"Decrypted Shutter transactions:{Environment.NewLine}{string.Join(Environment.NewLine, transactions.Select(tx => tx.ToShortString()))}");
+
+        Transaction[] filtered = FilterTransactions(transactions, parentHeader).ToArray();
+
+        ShutterTransactions shutterTransactions = new()
+        {
+            Transactions = filtered,
+            Slot = keys.Slot
+        };
+
+        Metrics.Transactions = (uint)filtered.Length;
+        Metrics.BadTransactions = (uint)(sequencedTransactions.Count - filtered.Length);
+
+        if (_logger.IsDebug && shutterTransactions.Transactions.Length > 0) _logger.Debug($"Filtered Shutter transactions:{Environment.NewLine}{string.Join(Environment.NewLine, shutterTransactions.Transactions.Select(tx => tx.ToShortString()))}");
+        return shutterTransactions;
+    }
+
+    public void LoadFromReceipts(Block head, TxReceipt[] receipts, ulong eon)
+    {
+        lock (_events)
+        {
+            if (_loadFromReceipts && head is not null)
+            {
+                var events = _logScanner.ScanReceipts(head.Number, receipts).ToList();
+                _events.EnqueueEvents(events, eon);
+            }
+        }
+    }
+
+    private IEnumerable<Transaction> FilterTransactions(IEnumerable<Transaction> transactions, BlockHeader parentHeader)
+    {
+        foreach (Transaction tx in transactions)
+        {
+            if (_txFilter.IsAllowed(tx, parentHeader) == TxPool.AcceptTxResult.Accepted)
+            {
+                yield return tx;
+            }
+        }
+    }
+
+    private Transaction[] DecryptSequencedTransactions(List<SequencedTransaction> sequencedTransactions, List<(byte[], byte[])> decryptionKeys)
+    {
+        int txCount = sequencedTransactions.Count;
+        int keyCount = decryptionKeys.Count;
+
+        if (txCount < keyCount)
+        {
+            _logger.Error($"Could not decrypt Shutter transactions: found {txCount} transactions but received {keyCount} keys (excluding placeholder).");
+            return [];
+        }
+
+        if (txCount > keyCount)
+        {
+            _logger.Warn($"Could not decrypt all Shutter transactions: found {txCount} transactions but received {keyCount} keys (excluding placeholder).");
+            sequencedTransactions = sequencedTransactions[..keyCount];
+            txCount = keyCount;
+        }
+
+        using ArrayPoolList<SequencedTransaction> sortedIndexes = sequencedTransactions.ToPooledList();
```

Comment:
if `sequencedTransactions` are already a list, you don't need to copy them to another list.

Reply:
I think you need to copy the list here, so you can sort the indexes but keep the original list in the same order

Reply:
You need to copy so you can sort and keep the original list in order. Alternatively you could just use one list and sort it back to the original order - you think that's better?